### DISCUSSION TITLE
How is AutomaticCFG different from the other CFG tools?

### DISCUSSION THREAD

**Discussion by TomLucidor** [Extraltodeus/ComfyUI-AutomaticCFG#53] [Q&A]
There is a whole list of CFG augmentation tools that seems to provide different results, do they tend to follow similar path to AutomaticCFG?
- Dynamic Thresholding ("X mimic Y clamp Z%") https://github.com/mcmonkeyprojects/sd-dynamic-thresholding
- CFG++ or "manifold-constrained" https://github.com/CFGpp-diffusion/CFGpp
- Smoothed Energy Guidance, Semantic CFG, Perturbed Attention Guidance, CFG Interval, CFG Scheduler https://github.com/v0xie/sd-webui-incantations
- Skimmed CFG (thanks bro) https://github.com/Extraltodeus/Skimmed_CFG
- CFG Rescale https://github.com/Seshelle/CFG_Rescale_webui https://github.com/MythicalChu/CFGRescale_For_Forge
- Adaptive Guidance https://github.com/asagi4/ComfyUI-Adaptive-Guidance
- APG Scaling https://github.com/logtd/ComfyUI-APGScaling
- Semantic CFG https://github.com/shiimizu/ComfyUI-semantic-aware-guidance

**Comment by bkutasi** [comment]
Let's add this to it as well:

- https://github.com/MythicalChu/ComfyUI-APG_ImYourCFGNow

But it's a valid question. I guess a lot of other nodes are retakes of various CFG scaling and guidance research.
The latest and most promising is adaptive projected guidance (APG). I would love to see it implemented in AutomaticCFG. As for differences, there are minor ones in details, but most of the time it's another way of getting the image.

**Comment by Extraltodeus** [comment]
And here I was having fun at writing yet another method lol

Well, Automatic CFG is more of a simple way of avoiding burns. The method itself is pretty simple: take 25% of the highest values of each channels and 25% of the lowest values as positive. Average them all as a single value which, as I noticed while testing the idea, falls near 0.8 when the images are "good". Then proceed by playing the CFG function, using a CFG scale rescaled according to this result and aiming for a resulting value being your sampler's CFG divided by then. Since the default is 8.

**Reply by TomLucidor** [reply]
Sounds weird, but was hoping for a method-by-method tear down of each of these algorithms (if you wish to do so), and what are the strengths/weaknesses of these major methods. Every method looks fun but are hard to understand in a sense.

**Reply by Extraltodeus** [reply]
The thing is that I know my own and not much of the others. I actually used your list to discover some 😂

**Reply by Extraltodeus** [reply]
I've got a few closed tickets on the skimmed cfg repository where I explained how it works (sincerely sorry, I'm on my phone right now)

**Reply by TomLucidor** [reply]
@Extraltodeus if my list is useful then hope ya likeit cus all the other new models need support as well
P.S. if only documentation and explainers are easy to handle, but kinda need to poke around you recs on the tickets first

**Reply by TomLucidor** [reply]
Note to self: have to read this report later https://github.com/v0xie/sd-webui-incantations/issues/63#issuecomment-2442599979
